### PR TITLE
feat: 1836 Add create charge WebApi

### PR DESCRIPTION
**Pull request by ShaBloze**
## Description

Add create charge in WebApi.
Create a ChargeInformationCommand and send it through ServiceBus.

## References

* #1836

### CODE REVIEW

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Controllers/V1/ChargesController.cs [hunk 1]
```diff
@@ -50,5 +56,19 @@ public async Task<IActionResult> SearchAsync(ChargeSearchCriteriaV1Dto searchCri
 
             return Ok(charges);
         }
+
+        /// <summary>
+        /// Sends a 'ChargeInformationCommand'
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Controllers/V1/ChargesController.cs:61]: "Sends"? I'd say it creates a ChargeInformationCommand for Charge information handling.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Factories/ChargeInformationCommandFactory.cs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+using Resolution = GreenEnergyHub.Charges.Domain.Charges.Resolution;
+using VatClassification = GreenEnergyHub.Charges.Domain.Charges.VatClassification;
+
+namespace GreenEnergyHub.Charges.WebApi.Factories;
+
+public class ChargeInformationCommandFactory : IChargeInformationCommandFactory
+{
+    private static IClock _clock;
+
+    private static string MeteringPointAdministratorGln => "5790001330552";
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Factories/ChargeInformationCommandFactory.cs:34]: I suggest we put this constant in a `MarketParticipantConstants` class in the same namespace as `MarketParticipantRole.cs`

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Factories/IChargeInformationCommandFactory.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Energinet.DataHub.Charges.Contracts.Charge;
+using GreenEnergyHub.Charges.Domain.Dtos.ChargeInformationCommands;
+
+namespace GreenEnergyHub.Charges.WebApi.Factories;
+
+public interface IChargeInformationCommandFactory
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Factories/IChargeInformationCommandFactory.cs:20]: Remember to document interfaces :-)

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Controllers/V1/ChargesController.cs [hunk 4]
```diff
@@ -50,5 +59,21 @@ public async Task<IActionResult> SearchAsync(ChargeSearchCriteriaV1Dto searchCri
...
+
+        /// <summary>
+        /// Creates a 'ChargeInformationCommand' for charge information handling.
+        /// </summary>
+        /// <returns>"200 OK"</returns>
+        [HttpPost("CreateAsync")]
+        [MapToApiVersion(Version1)]
+        public async Task<IActionResult> CreateAsync([FromBody] CreateChargeInformationV1Dto chargeInformation)
+        {
+            _correlationContext.SetId(Guid.NewGuid().ToString());
```
Comment (ShaBloze) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Controllers/V1/ChargesController.cs:71]: Unsure if this should be here or be in a middleware or something else?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/WebApi/Factory/ChargeInformationCommandFactoryTests.cs [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+using Resolution = GreenEnergyHub.Charges.Domain.Charges.Resolution;
+
+namespace GreenEnergyHub.Charges.Tests.WebApi.Factory
+{
+    [UnitTest]
+    public class ChargeInformationCommandFactoryTests
+    {
+        [Theory]
+        [InlineAutoData]
+        public void Create_ReturnsChargeInformationCommand_BasedOnCreateChargeInformationV1Dto(
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/WebApi/Factory/ChargeInformationCommandFactoryTests.cs:38]:
```suggestion
        public void Create_CreateChargeInformationV1Dto_ReturnsChargeInformationCommand(
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Factories/ChargeInformationCommandFactory.cs [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    private DocumentDto CreateDocument(CreateChargeInformationV1Dto chargeInformation)
+    {
+        var documentId = "dh" + Guid.NewGuid().ToString("N");
+
+        var document = new DocumentDto(
+            documentId,
+            _clock.GetCurrentInstant(),
+            DocumentType.RejectRequestChangeOfPriceList,
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Factories/ChargeInformationCommandFactory.cs:54]: This should be 
```suggestion
            DocumentType.RequestChangeOfPriceList,
```

File: source/Contracts/Charge/CreateChargeInformationV1Dto.cs [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+
+// ReSharper disable once CheckNamespace - Type is shared so namespace is not determined by project structure/namespace
+namespace Energinet.DataHub.Charges.Contracts.Charge
+{
+    public record CreateChargeInformationV1Dto(
```
Comment (ShaBloze) [inline source/Contracts/Charge/CreateChargeInformationV1Dto.cs:20]: Should this be renamed to `CreateChargeV1Dto` now that the Client, WebApi etc. is called `CreateCharge` ?

Comment (prtandrup) [reply source/Contracts/Charge/CreateChargeInformationV1Dto.cs:20]: Let's do that.

Comment (HenrikSommer) [reply source/Contracts/Charge/CreateChargeInformationV1Dto.cs:20]: Yes, outside the Charges domain there is no knowledge of anything called `ChargeInformation`.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/SharedDtos/MarketParticipantConstants.cs [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace GreenEnergyHub.Charges.Domain.Dtos.SharedDtos
+{
+    public static class MarketParticipantConstants
+    {
+        public static string MeteringPointAdministratorGln => "5790001330552";
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/SharedDtos/MarketParticipantConstants.cs:19]: This was probably made due to one of my comments - but thinking again, this GLN identifies the danish DataHub, which should not be a constant value in code that could someday become _Green Energy Hub_ code. We should probably get the value from the configuration.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/WebApi/Factory/ChargeInformationCommandFactoryTests.cs [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+            actualDocument.Sender.BusinessProcessRole.Should()
+                .Be(MarketParticipantRoleMapper.Map(senderMarketParticipant.BusinessProcessRole));
+            actualDocument.Sender.B2CActorId.Should().Be(Guid.Empty);
+            actualDocument.Recipient.BusinessProcessRole.Should().Be(MarketParticipantRole.MeteringPointAdministrator);
+            actualDocument.Recipient.MarketParticipantId.Should().Be(MarketParticipantConstants.MeteringPointAdministratorGln);
+            actualDocument.Recipient.B2CActorId.Should().Be(Guid.Empty);
+            actualDocument.IndustryClassification.Should().Be(IndustryClassification.Electricity);
+            actualDocument.BusinessReasonCode.Should().Be(BusinessReasonCode.UpdateChargeInformation);
+
+            var actualOperation = actual.Operations.First();
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/WebApi/Factory/ChargeInformationCommandFactoryTests.cs:74]:
```suggestion
            var actualOperation = actual.Operations.Single();
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/EnvironmentSettingNames.cs [hunk 10]
```diff
@@ -32,5 +32,8 @@ public static class EnvironmentSettingNames
 
         // Localization
         public const string LocalTimeZoneName = "LOCAL_TIMEZONENAME";
+
+        public const string ChargesDomainEventTopicName = "DOMAIN_EVENTS_TOPIC_NAME";
+        public const string DataHubSenderConnectionString = "INTEGRATIONEVENT_SENDER_CONNECTION_STRING";
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/EnvironmentSettingNames.cs:37]:
```suggestion
        public const string DataHubSenderConnectionString = "DOMAINEVENT_SENDER_CONNECTION_STRING";
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTest.Core/Fixtures/WebApi/ChargesWebApiFixture.cs [hunk 11]
```diff
@@ -54,11 +68,20 @@ protected override async Task OnInitializeWebApiDependenciesAsync(IConfiguration
             Environment.SetEnvironmentVariable(EnvironmentSettingNames.FrontEndOpenIdUrl, AuthorizationConfiguration.FrontendOpenIdConfigurationUrl);
             Environment.SetEnvironmentVariable(EnvironmentSettingNames.FrontEndServiceAppId, AuthorizationConfiguration.FrontendApp.AppId);
             Environment.SetEnvironmentVariable(EnvironmentSettingNames.LocalTimeZoneName, "Europe/Copenhagen");
+
+            Environment.SetEnvironmentVariable(EnvironmentSettingNames.DataHubSenderConnectionString, ServiceBusResourceProvider.ConnectionString);
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTest.Core/Fixtures/WebApi/ChargesWebApiFixture.cs:72]:
```suggestion
            Environment.SetEnvironmentVariable(EnvironmentSettingNames.DomainEventSenderConnectionString, ServiceBusResourceProvider.ConnectionString);
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/QueryApiConfiguration.cs [hunk 12]
```diff
@@ -38,11 +50,35 @@ public static IServiceCollection AddQueryApi(this IServiceCollection serviceColl
                     options.UseSqlServer(connectionString);
                 });
 
+            serviceCollection.AddSingleton(_ =>
+            {
+                var connectionString = configuration.GetValue<string>(EnvironmentSettingNames.DataHubSenderConnectionString);
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/QueryApiConfiguration.cs:55]:
```suggestion
                var connectionString = configuration.GetValue<string>(EnvironmentSettingNames.DomainEventSenderConnectionString);
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Factories/ChargeInformationCommandFactory.cs [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+    {
+        var operations = CreateOperation(chargeInformation);
+        var document = CreateDocument(chargeInformation);
+
+        return new ChargeInformationCommand(document, operations);
+    }
+
+    private DocumentDto CreateDocument(CreateChargeInformationV1Dto chargeInformation)
+    {
+        var documentId = "dh" + Guid.NewGuid().ToString("N");
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Factories/ChargeInformationCommandFactory.cs:49]:
```suggestion
        var documentId = $"geh{Guid.NewGuid():N}";
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Factories/ChargeInformationCommandFactory.cs [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+
+        var document = new DocumentDto(
+            documentId,
+            _clock.GetCurrentInstant(),
+            DocumentType.RequestChangeOfPriceList,
+            _clock.GetCurrentInstant(),
+            new MarketParticipantDto(
+                Guid.NewGuid(),
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Factories/ChargeInformationCommandFactory.cs:57]: Maybe it is better to set a _Guid.Empty_, so no one thinks it is the id of an actual MarketParticipant. Same with the receiver.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Factories/ChargeInformationCommandFactory.cs [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+                Guid.Empty),
+            IndustryClassification.Electricity,
+            BusinessReasonCode.UpdateChargeInformation);
+
+        return document;
+    }
+
+    private static IReadOnlyCollection<ChargeInformationOperationDto> CreateOperation(CreateChargeInformationV1Dto chargeInformation)
+    {
+        var operationId = Guid.NewGuid().ToString("N");
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Factories/ChargeInformationCommandFactory.cs:74]:
```suggestion
        var operationId = $"{Guid.NewGuid():N}";
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/appsettings.Development.json.sample [hunk 16]
```diff
@@ -4,5 +4,8 @@
   },
   "FRONTEND_OPEN_ID_URL": "FRONTEND_OPEN_ID_URL",
   "FRONTEND_SERVICE_APP_ID": "FRONTEND_SERVICE_APP_ID",
-  "LOCAL_TIMEZONENAME": "Europe/Copenhagen"
+  "LOCAL_TIMEZONENAME": "Europe/Copenhagen",
+
+  "DOMAIN_EVENTS_TOPIC_NAME": "DOMAIN_EVENTS_TOPIC_NAME",
+  "INTEGRATIONEVENT_SENDER_CONNECTION_STRING": "INTEGRATIONEVENT_SENDER_CONNECTION_STRING"
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/appsettings.Development.json.sample:10]:
```suggestion
  "DOMAINEVENT_SENDER_CONNECTION_STRING": "DOMAINEVENT_SENDER_CONNECTION_STRING"
```

File: source/Energinet.DataHub.Charges.Libraries/source/Energinet.DataHub.Charges.Clients/Charges/ChargesClient.cs [hunk 17]
```diff
@@ -112,6 +112,22 @@ public async Task<ChargePricesV1Dto> SearchChargePricesAsync(ChargePricesSearchC
             return await HandleResultAsync<ChargePricesV1Dto>(response).ConfigureAwait(false);
         }
 
+        /// <summary>
+        /// Creates a 'ChargeInformationCommand' for charge information handling.
+        /// </summary>
+        public async Task CreateChargeAsync(CreateChargeV1Dto createChargeInformation)
+        {
+            var response = await _httpClient
+                .PostAsJsonAsync(ChargesRelativeUris.CreateCharge(), createChargeInformation)
```
Comment (HenrikSommer) [inline source/Energinet.DataHub.Charges.Libraries/source/Energinet.DataHub.Charges.Clients/Charges/ChargesClient.cs:121]:
```suggestion
        public async Task CreateChargeAsync(CreateChargeV1Dto createChargeV1Dto)
        {
            var response = await _httpClient
                .PostAsJsonAsync(ChargesRelativeUris.CreateCharge(), createChargeV1Dto)
```

File: source/Energinet.DataHub.Charges.Libraries/source/Energinet.DataHub.Charges.Clients/Charges/ChargesClient.cs [hunk 18]
```diff
@@ -112,6 +112,22 @@ public async Task<ChargePricesV1Dto> SearchChargePricesAsync(ChargePricesSearchC
...
+                .ConfigureAwait(false);
```
Comment (HenrikSommer) [inline source/Energinet.DataHub.Charges.Libraries/source/Energinet.DataHub.Charges.Clients/Charges/ChargesClient.cs:122]:
```suggestion
        public async Task CreateChargeAsync(CreateChargeV1Dto createChargeV1Dto)
        {
            var response = await _httpClient
                .PostAsJsonAsync(ChargesRelativeUris.CreateCharge(), createChargeV1Dto)
                .ConfigureAwait(false);
```

File: source/Energinet.DataHub.Charges.Libraries/source/Energinet.DataHub.Charges.Clients/Charges/IChargesClient.cs [hunk 19]
```diff
@@ -56,5 +56,10 @@ public interface IChargesClient
         /// </summary>
         /// <returns>Charge messages(Dto) with collection of messages</returns>
         Task<ChargeMessagesV1Dto> SearchChargeMessagesAsync(ChargeMessagesSearchCriteriaV1Dto searchCriteria);
+
+        /// <summary>
+        /// Creates a 'ChargeInformationCommand' for charge information handling.
+        /// </summary>
+        Task CreateChargeAsync(CreateChargeV1Dto createChargeInformation);
```
Comment (HenrikSommer) [inline source/Energinet.DataHub.Charges.Libraries/source/Energinet.DataHub.Charges.Clients/Charges/IChargesClient.cs:63]:
```suggestion
        Task CreateChargeAsync(CreateChargeV1Dto createChargeV1Dto);
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/WebApi/Factory/ChargeInformationCommandFactoryTests.cs [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+using Resolution = GreenEnergyHub.Charges.Domain.Charges.Resolution;
+
+namespace GreenEnergyHub.Charges.Tests.WebApi.Factory
+{
+    [UnitTest]
+    public class ChargeInformationCommandFactoryTests
+    {
+        [Theory]
+        [InlineAutoData]
+        public void Create_CreateChargeInformationV1Dto_ReturnsChargeInformationCommand(
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/WebApi/Factory/ChargeInformationCommandFactoryTests.cs:38]:
```suggestion
        public void Create_createChargeV1Dto_ReturnsChargeInformationCommand(
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Factories/IChargeInformationCommandFactory.cs [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+/// Factory for creating <see cref="ChargeInformationCommand"/>
+/// </summary>
+public interface IChargeInformationCommandFactory
+{
+    /// <summary>
+    /// Factory for creating <see cref="ChargeInformationCommand"/> from <see cref="CreateChargeV1Dto"/>
+    /// </summary>
+    /// <param name="charge"></param>
+    /// <returns>A <see cref="ChargeInformationCommand"/></returns>
+    ChargeInformationCommand Create(CreateChargeV1Dto charge);
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Factories/IChargeInformationCommandFactory.cs:30]:
```suggestion
    /// <param name="createChargeV1Dto"></param>
    /// <returns>A <see cref="ChargeInformationCommand"/></returns>
    ChargeInformationCommand Create(CreateChargeV1Dto createChargeV1Dto);
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Factories/ChargeInformationCommandFactory.cs [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+        _meteringPointAdministratorGln = meteringPointAdministratorGln;
+    }
+
+    public ChargeInformationCommand Create(CreateChargeV1Dto charge)
+    {
+        var operations = CreateOperation(charge);
+        var document = CreateDocument(charge);
+
+        return new ChargeInformationCommand(document, operations);
+    }
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Factories/ChargeInformationCommandFactory.cs:47]:
```suggestion
    public ChargeInformationCommand Create(CreateChargeV1Dto createChargeV1Dto)
    {
        var operations = CreateOperation(createChargeV1Dto);
        var document = CreateDocument(createChargeV1Dto);

        return new ChargeInformationCommand(document, operations);
    }
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/appsettings.Development.json.sample [hunk 23]
```diff
@@ -4,5 +4,9 @@
   },
   "FRONTEND_OPEN_ID_URL": "FRONTEND_OPEN_ID_URL",
   "FRONTEND_SERVICE_APP_ID": "FRONTEND_SERVICE_APP_ID",
-  "LOCAL_TIMEZONENAME": "Europe/Copenhagen"
+  "LOCAL_TIMEZONENAME": "Europe/Copenhagen",
+
+  "DOMAIN_EVENTS_TOPIC_NAME": "DOMAIN_EVENTS_TOPIC_NAME",
+  "INTEGRATIONEVENT_SENDER_CONNECTION_STRING": "DOMAINEVENT_SENDER_CONNECTION_STRING",
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/appsettings.Development.json.sample:10]:
```suggestion
  "INTEGRATIONEVENT_SENDER_CONNECTION_STRING": "INTEGRATIONEVENT_SENDER_CONNECTION_STRING",
```

**Review by prtandrup — APPROVED** [review]
LGTM

**Review by dstenroejl — APPROVED** [review]

**Pull request merged**